KlugerLab/GeneTrajectory-python
# Release v1.0

Creating release
- changed package name to `gene_trajectory`
- added `compute_gene_distance_cmd.py` to run EMD computations from command line (used in the R package)

## Testing ##
The simplest way to use the Python functions from R and to check whether the environment is properly set up 
- create a virtual environment with the 'gene_trajectory' package if it's not existing
- export `cal_ot_mat_from_numpy` to a R function
- use the function using input matrices
```
if(!reticulate::virtualenv_exists('gene_trajectory')){
  reticulate::virtualenv_create('gene_trajectory', packages=c('gene_trajectory'))
}
reticulate::use_virtualenv('gene_trajectory')

cal_ot_mat_from_numpy <-reticulate::import('gene_trajectory.compute_gene_distance_cmd')$cal_ot_mat_from_numpy

gdm <- matrix(c(0, 1, 2, 1, 0, 2, 2, 2, 1), nrow = 3)
gem <- matrix(c(.3, .1, .6, .2, .3, .5,.6, .2, .2), nrow = 3)
exp <- matrix(c(0.0, 0.8, 1.0, 0.8, 0.0, 0.9, 1.0, 0.9, 0.0), nrow = 3)
cal_ot_mat_from_numpy(ot_cost = gdm, gene_expr = gem)
```


## Installation ##
The easiest way is to create a virtualenv for gene_trajectory using [reticulate](https://rstudio.github.io/reticulate/index.html)
```
if(!reticulate::virtualenv_exists('gene_trajectory')){
  reticulate::virtualenv_create('gene_trajectory', packages=c('gene_trajectory'))
}
reticulate::use_virtualenv('gene_trajectory')
```
or to add to an existing virtualenv using
```
reticulate::py_install("gene-trajectory")
```

In general (especially in a conda environment) it can be installed with pip as 
```
system(sprintf('%s -m pip install gene-trajectory', reticulate::py_exe()))
```

The development version can be installed as
```
system(sprintf('%s -m pip install git+https://github.com/Klugerlab/GeneTrajectory-python.git', reticulate::py_exe()))
```
This works both on virtualenv and conda.

## Running cal_ot_mat from a path where `ot_cost.csv` and `gene_expression.mtx` are present #
The easiest way to achieve this from R is to create a function `cal_ot_mat` that takes the path to a folder with the input files. 
```
cal_ot_mat <- reticulate::import('gene_trajectory.compute_gene_distance_cmd')$cal_ot_mat
cal_ot_mat('data/test')
```

It's also possible to run this as a command from  the command line as
```
python  -m gene_trajectory.compute_gene_distance_cmd --path data/test
```
where python may be changed with a specific python executable with GeneTrajectory-python installed (e.g. it could be in a virtual environment as `~/.virtualenvs/gene_trajectory/bin/python`)

The same command can be run as a process from R as 
```
system(sprintf('%s -m gene_trajectory.compute_gene_distance_cmd --path %s', reticulate::py_exe(), 'data/test'))
```

## Running cal_ot_mat using R matrices ##
Assuming the reticulate environment has gene_trajectory installed, one can import to a R function called e.g.
`cal_ot_mat_from_numpy` which accepts matrices and automatically convert them to numpy arrays as
```
cal_ot_mat_from_numpy <- reticulate::import('gene_trajectory.compute_gene_distance_cmd')$cal_ot_mat_from_numpy

gdm <- matrix(c(0, 1, 2, 1, 0, 2, 2, 2, 1), nrow = 3)
gem <- matrix(c(.3, .1, .6, .2, .3, .5,.6, .2, .2), nrow = 3)
exp <- matrix(c(0.0, 0.8, 1.0, 0.8, 0.0, 0.9, 1.0, 0.9, 0.0), nrow = 3)
cal_ot_mat_from_numpy(ot_cost = gdm, gene_expr = gem)
```
which returns
```
100%|██████████| 3/3 [00:00<00:00, 138.71it/s]
     [,1] [,2] [,3]
[1,]  0.0  0.8  1.0
[2,]  0.8  0.0  0.9
[3,]  1.0  0.9  0.0
```
